### PR TITLE
Added redirectURL to validation mail

### PR DESCRIPTION
**Pull request by metodej**
- url in validation mail now contains redirectURL
- redirectURL added to fedInfo in this PR: [https://github.com/CESNET/perun-wui/pull/148](url)

### CODE REVIEW

File: perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/MailManagerImpl.java [hunk 1]
```diff
@@ -1729,6 +1729,7 @@ private void mailValidation(Application app, ApplicationMail mail, List<Applicat
 									newValue += "?vo="+ getUrlEncodedString(app.getVo().getShortName());
 									newValue += ((app.getGroup() != null) ? "&group="+ getUrlEncodedString(app.getGroup().getName()) : EMPTY_STRING);
 									newValue += "&i=" + URLEncoder.encode(i, StandardCharsets.UTF_8) + "&m=" + URLEncoder.encode(m, StandardCharsets.UTF_8);
+									newValue += "&target=" + BeansUtils.stringToMapOfAttributes(app.getFedInfo()).get("redirectURL");
```
Comment (Vojtech-Sassmann) [inline perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/MailManagerImpl.java:1732]: We probably shouldn't add the target if the redirectURL is not filled.

**Review by Vojtech-Sassmann — CHANGES_REQUESTED** [review]

File: perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/MailManagerImpl.java (hunk 1 above)
Comment (metodej) [reply perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/MailManagerImpl.java]: Fixed, thanks.

**Review by Vojtech-Sassmann — CHANGES_REQUESTED** [review]
In method `processApplication` in `RegistrarManagerImpl` there is this code: 
```
		LinkedHashMap<String, String> map = new LinkedHashMap<>(session.getPerunPrincipal().getAdditionalInformations());
		String additionalAttrs = BeansUtils.attributeValueToString(map, LinkedHashMap.class.getName());
		application.setFedInfo(additionalAttrs);
```

This rewrites the fedInfo send from the gui. Therefore, the URL is not saved in DB. You can probably parse the URL and set it to the `additionalAttrs` map.

**Comment by metodej** [comment]
> 
> 
> In method `processApplication` in `RegistrarManagerImpl` there is this code:
> 
> ```java
> 		LinkedHashMap<String, String> map = new LinkedHashMap<>(session.getPerunPrincipal().getAdditionalInformations());
> 		String additionalAttrs = BeansUtils.attributeValueToString(map, LinkedHashMap.class.getName());
> 		application.setFedInfo(additionalAttrs);
> ```
> 
> This rewrites the fedInfo send from the gui. Therefore, the URL is not saved in DB. You can probably parse the URL and set it to the `additionalAttrs` map.

Ok, I've added it to `additionalAttrs`, so it should be saved in URL now.

File: perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/RegistrarManagerImpl.java [hunk 2]
```diff
@@ -3925,8 +3925,10 @@ private int processApplication(PerunSession session, Application application, Li
 			}
 		}
 
-		// store user-ext-source attributes from session to application object
+		// store user-ext-source attributes and redirectURL from session to application object
 		LinkedHashMap<String, String> map = new LinkedHashMap<>(session.getPerunPrincipal().getAdditionalInformations());
+		String redirectURL = BeansUtils.stringToMapOfAttributes(application.getFedInfo()).get("redirectURL");
```
Comment (Vojtech-Sassmann) [inline perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/RegistrarManagerImpl.java:3930]: This is wrong. The method `stringToMapOfAttributes` is used for a completely different purpose. 

The fedInfo has a JSON format, something like this: 
```
{ displayName="Vojta Petr von Sassmann" commonName="null" givenName="Vojta" sureName="von Sassmann" loa="null" mail="sdf@sdf.sd" organization="true" redirectURL="localhost/krb/registrar/?group=group" }
```

From such a string, you need to parse the redirectURL. You can probably use some Json parsing library and use it for this purpose. 

Also, please make these changes as soon as possible, we need this functionality to be finished by the end of this week.

**Review by Vojtech-Sassmann — CHANGES_REQUESTED** [review]

File: perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/MailManagerImpl.java [hunk 3]
```diff
@@ -1729,6 +1729,8 @@ private void mailValidation(Application app, ApplicationMail mail, List<Applicat
 									newValue += "?vo="+ getUrlEncodedString(app.getVo().getShortName());
 									newValue += ((app.getGroup() != null) ? "&group="+ getUrlEncodedString(app.getGroup().getName()) : EMPTY_STRING);
 									newValue += "&i=" + URLEncoder.encode(i, StandardCharsets.UTF_8) + "&m=" + URLEncoder.encode(m, StandardCharsets.UTF_8);
+									String redirectURL = BeansUtils.stringToMapOfAttributes(app.getFedInfo()).get("redirectURL");
+									newValue += (redirectURL != null) ? "&target=" + redirectURL : EMPTY_STRING;
```
Comment (Vojtech-Sassmann) [inline perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/MailManagerImpl.java:1733]: The notification template can contain either `{validationLink-[krb|non|...|]}` or `{validationLink}`. This is handling only the first case. For the second case, please add this as well. It is a couple of lines below.

**Review by Vojtech-Sassmann — CHANGES_REQUESTED** [review]

**Review by Vojtech-Sassmann — APPROVED** [review]

**Pull request merged**